File: src/main/kotlin/web/snapshots/FileSystemPathProvider.kt
package web.snapshots

import digests.SHA224
import java.net.URL

class FileSystemPathProvider {
    private val urlPrefixAndSuffixRegex = Regex("^https?://(www\\.)?|[?#].+$", RegexOption.IGNORE_CASE)
    private val fsUnfriendlyCharactersRegex = Regex("[^a-z0-9_.-]+", RegexOption.IGNORE_CASE)

    fun getSnapshotDirPath(url: String): String {
        val hostName = URL(url).host.removePrefix("www.")

        val urlFsSafe = toFsSafeString(url)
            .removePrefix(hostName)
            .trimStart('_')

        val firstLetter = "${hostName}_".uppercase()[0]
        val optionalDash = if ("" == urlFsSafe) "" else "-"
        val urlHash = SHA224.hexOf(url)

        return "$firstLetter/$hostName/$urlFsSafe$optionalDash$urlHash"
    }

    private fun toFsSafeString(url: String): String {
        return url
            .replace(urlPrefixAndSuffixRegex, "")
            .replace(fsUnfriendlyCharactersRegex, "_")
            .trim('_')
    }
}

File: symfony/src/Utils/Web/Snapshots/FileSystemPathProvider.php
<?php

declare(strict_types=1);

namespace App\Utils\Web\Snapshots;

use App\Utils\Web\Url\Url;
use Psl\Str;
use TRegx\CleanRegex\Pattern;

class FileSystemPathProvider
{
    private readonly Pattern $urlPrefixAndSuffixRegex;
    private readonly Pattern $fsUnfriendlyCharactersRegex;

    public function __construct()
    {
        $this->urlPrefixAndSuffixRegex = Pattern::of('^https?://(www\.)?|[?#].+$', 'i');
        $this->fsUnfriendlyCharactersRegex = Pattern::of('[^a-z0-9_.-]+', 'i');
    }

    public function getSnapshotDirPath(Url $url): string
    {
        $hostName = Str\strip_prefix($url->getHost(), 'www.');

        $urlFsSafe = ltrim(Str\strip_prefix($this->toFileSystemSafeString($url->getUrl()), $hostName), '_');

        $firstLetter = mb_strtoupper("{$hostName}_")[0];
        $optionalDash = '' === $urlFsSafe ? '' : '-';
        $urlHash = hash('sha224', $url->getUrl());

        return "$firstLetter/$hostName/$urlFsSafe$optionalDash$urlHash";
    }

    private function toFileSystemSafeString(string $url): string
    {
        $url = $this->urlPrefixAndSuffixRegex->prune($url);
        $url = $this->fsUnfriendlyCharactersRegex->replace($url)->with('_');

        return trim($url, '_');
    }
}
